test/performance/perf-upsert/src/main.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Performance test for Materialize Upsert sources
+
+#![deny(missing_debug_implementations, missing_docs)]
+
+use std::process;
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::Result;
+use rand::Rng;
+use structopt::StructOpt;
+
+use test_util::kafka::kafka_client;
+
+use crate::config::{Args, KafkaConfig, MzConfig};
+use crate::mz_client::MzClient;
+
+mod config;
+mod mz_client;
+
+#[tokio::main]
+async fn main() {
+    if let Err(e) = run().await {
+        println!("ERROR: {}", e);
+        let mut err = e.source();
+        while let Some(e) = err {
+            println!("    caused by: {}", e);
+            err = e.source();
+        }
```

Comment:
Do lifetime issues prevent this from being:

```suggestion
        while let Some(e) = e.source() {
            println!("    caused by: {}", e);
        }
```

Reply:
this turns into an infinite loop -- i think because e is bound to e.source within the let statement but the binding doesnt persist

Reply:
oops no I wasn't doing the right thing in testing -- its actually kind of hard to trigger an error that makes this test quit early :/ im going to punt on this for now to push this pr out but I'll add a todo

Reply:
👍 just checking